File: Notes/Computer Science/Algorithms/Meta/Sebastien Rubens Meta Puzzles Solutions/go/l1_cafeteria/l1_cafeteria.go
// meta_puzzles by Sebastien Rubens
//
// Please go to https://github.com/seb-pg/meta_puzzles/README.md
// for more information
//
// To the extent possible under law, the person who associated CC0 with
// meta_puzzles has waived all copyright and related or neighboring rights
// to meta_puzzles.
//
// You should have received a copy of the CC0 legalcode along with this
// work.  If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.

package l1_cafeteria

import (
	"meta_puzzles/test"
	"sort"
)

func getMaxAdditionalDinersCount(N int64, K int64, M int32, S []int64) int64 {
	d := K + 1
	taken := make([]int64, 0, len(S)+2)
	taken = append(taken, 0-K)
	taken = append(taken, S...)
	taken = append(taken, N+d)
	sort.Slice(taken, func(i, j int) bool { return taken[i] < taken[j] })

	var nb int64 = 0
	e := len(taken) - 1
	for i := 0; i < e; i += 1 {
		nb += (taken[i+1] - taken[i] - d) / d
	}
	return nb
}

type Args struct {
	N   int64
	K   int64
	S   []int64
	res int64
}

func (self Args) GetResult() any {
	return self.res
}

func Tests() uint {
	wrapper := func(p Args) int64 { return getMaxAdditionalDinersCount(p.N, p.K, int32(len(p.S)), p.S) }

	args_lists := []Args{
		{10, 1, []int64{2, 6}, 3},
		{15, 2, []int64{11, 6, 14}, 1},
	}

	return test.RunAllTests("l1_cafetaria", args_lists, wrapper)
}

File: Notes/Computer Science/Algorithms/Meta/Sebastien Rubens Meta Puzzles Solutions/cpp17/l1_cafeteria.cpp
// meta_puzzles by Sebastien Rubens
//
// Please go to https://github.com/seb-pg/meta_puzzles/README.md
// for more information
//
// To the extent possible under law, the person who associated CC0 with
// meta_puzzles has waived all copyright and related or neighboring rights
// to meta_puzzles.
//
// You should have received a copy of the CC0 legalcode along with this
// work.  If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.

#include "test_all.h"

#include <algorithm>
#include <cstdint>  // int**_t
#include <vector>

namespace l1_cafeteria {

int64_t getMaxAdditionalDinersCountCpp17(int64_t N, int64_t K, uint32_t M, const std::vector<long long>& S)
{
    // https://www.metacareers.com/profile/coding_puzzles/?puzzle=203188678289677
    // Constraints
    //      1 ≤ N ≤ 10^15       N is the number of seats
    //      1 ≤ K ≤ N           K is the number of empty seats needed between occupied seats
    //      1 ≤ M ≤ 500,000     M is the number of diners
    //      1 ≤ Si ≤ N          Si is a seat
    // Complexity: O(M*log(M)), but the complexity could be O(M) if S was sorted

    (void)M;

    const auto d = K + 1;
    std::vector<int64_t> taken;
    taken.reserve(S.size() + 2);
    taken.emplace_back(0ull - K);  // we are adding "fake" seats at the beginning
    taken.insert(std::end(taken), std::cbegin(S), std::cend(S));
    taken.insert(std::end(taken), N + d);  // we are adding "fake" seats at the end

    // we sort elements of S: O(M * log(M))
    std::sort(std::begin(taken) + 1, std::end(taken) - 1);

    // we are calculating the extra seats available between each consecutive seats: O(M)
    int64_t nb = 0;
    for (size_t i = 0, e = taken.size() - 1; i < e; ++i)
        nb += (taken[i + 1] - taken[i] - d) / d;
    return nb;  // result should always be positive
}

using namespace std;

long long getMaxAdditionalDinersCount(long long N, long long K, int M, vector<long long> S) {
    return getMaxAdditionalDinersCountCpp17(N, K, static_cast<uint32_t>(M), S);
}


struct Args
{
    long long N;
    long long K;
    std::vector<long long> S;  // error: could not convert ‘p.l1_cafeteria::Args::S’ from ‘vector<long int>’ to ‘vector<long long int>’
};

auto tests()
{
    const auto wrapper = [](Args& p)
    {
        return getMaxAdditionalDinersCount(p.N, p.K, static_cast<int>(p.S.size()), p.S);
    };

    std::vector<NamedTests<Args, long long>> tests = {
        { "Meta", {
                { { 10, 1, { 2, 6 } }, 3 },
                { { 15, 2, { 11, 6, 14 } }, 1 },
            }
        },
    };

    return run_all_tests("l1_cafetaria", tests, wrapper);
}

}  // namespace l1_cafetaria
